File: android/app/src/main/java/org/onemindlabs/airhop/tor/AirhopIPtProxy.kt
// The pluggable transports Arti dials to reach a bridge, built from
// native/iptproxy.
//
// Arti would normally run obfs4 and Snowflake as child processes, which iOS
// forbids, so both platforms run them in-process behind a SOCKS5 listener each.
// This owns their lifecycle and reports the ports they landed on.
//
// Mirrors ios/Airhop/AirhopIPtProxy.swift.
package org.onemindlabs.airhop.tor

import IPtProxy.Controller
import android.content.Context
import android.util.Log
import java.io.File

enum class AirhopTransport(val id: String) {
    OBFS4("obfs4"),
    SNOWFLAKE("snowflake"),
    ;

    companion object {
        // The transports named by a newline-separated list of bridge lines.
        //
        // Reads the first word of each line, where the Tor bridge format puts
        // the transport name. Not a parser: Arti parses the lines
        // authoritatively in start, so a line this misreads is refused there
        // because its transport's port will be 0.
        fun namedIn(bridgeLines: String): Set<AirhopTransport> {
            val found = mutableSetOf<AirhopTransport>()
            for (raw in bridgeLines.lineSequence()) {
                val line = raw.trim()
                if (line.isEmpty() || line.startsWith("#")) continue
                val words = line.split(" ").filter { it.isNotEmpty() }.toMutableList()
                // The word `Bridge` may lead the line; the transport follows it.
                if (words.firstOrNull()?.lowercase() == "bridge") words.removeAt(0)
                when (words.firstOrNull()) {
                    OBFS4.id -> found.add(OBFS4)
                    SNOWFLAKE.id -> found.add(SNOWFLAKE)
                }
            }
            return found
        }
    }
}

object AirhopIPtProxy {

    private const val TAG = "AirhopIPtProxy"

    // Beside Arti's directory rather than inside it: the panic wipe deletes
    // both, and Arti owns the layout of its own.
    private const val STATE_DIR = "iptproxy"

    // Null until something asks for a transport: the controller writes to its
    // state directory on construction, so it is not built for a user who never
    // turns bridges on.
    private var controller: Controller? = null
    private val running = mutableSetOf<AirhopTransport>()

    fun stateDir(context: Context): File = File(context.filesDir, STATE_DIR)

    // Start [transports], returning the loopback port each landed on, or null if
    // any could not start.
    //
    // A null is fatal to the caller: Arti would otherwise be asked for a bridge
    // whose transport is missing, and a user who asked for a bridge is likely
    // somewhere a direct connection is unsafe.
    @Synchronized
    fun start(context: Context, transports: Set<AirhopTransport>): Map<AirhopTransport, Int>? {
        if (transports.isEmpty()) return emptyMap()
        val controller = ensureController(context) ?: return null

        val ports = mutableMapOf<AirhopTransport, Int>()
        for (transport in transports) {
            if (!running.contains(transport)) {
                try {
                    // No upstream proxy: Arti is the only thing dialling these.
                    controller.start(transport.id, "")
                } catch (e: Throwable) {
                    // Throwable: a libgojni load or link failure arrives as an
                    // Error, and uncaught on this worker it kills the process.
                    Log.e(TAG, "${transport.id} did not start: ${e.message}")
                    stopLocked()
                    return null
                }
                running.add(transport)
            }
            // Ports are assigned by the library, never chosen. A zero means the
            // listener is not up despite start having returned.
            val port = controller.port(transport.id).toInt()
            if (port !in 1..65535) {
                Log.i(TAG, "${transport.id} reported port $port")
                stopLocked()
                return null
            }
            ports[transport] = port
        }
        return ports
    }

    @Synchronized
    fun stop() = stopLocked()

    // Panic wipe only: the state directory records which transports ran.
    @Synchronized
    fun wipeState(context: Context) {
        stopLocked()
        controller = null
        if (!stateDir(context).deleteRecursively()) {
            Log.w(TAG, "transport state directory was not fully removed")
        }
    }

    private fun stopLocked() {
        val controller = this.controller ?: return
        for (transport in running) {
            // One that throws must not strand the others: a half-done stop leaves
            // a transport this object thinks is running, and the next start would
            // skip it and hand Arti a stale port.
            try {
                controller.stop(transport.id)
            } catch (e: Throwable) {
                Log.w(TAG, "${transport.id} did not stop: ${e.message}")
            }
        }
        running.clear()
    }

    private fun ensureController(context: Context): Controller? {
        controller?.let { return it }

        val dir = stateDir(context)
        dir.mkdirs()

        // Logging off. The log records bridge addresses, which is the most
        // incriminating thing this device could write down, and nothing reads it.
        //
        // First line to touch the generated binding, so gomobile's static
        // initialiser loads libgojni here. A Go-side nil surfaces as a throw
        // rather than null, and a missing or rejected library as an Error, which
        // is why this catches Throwable. Either has to stop the start rather
        // than reach Arti.
        val created = try {
            Controller(dir.absolutePath, false, false, "ERROR", null)
        } catch (e: Throwable) {
            Log.e(TAG, "controller did not initialise: ${e.message}")
            return null
        }

        controller = created
        return created
    }
}

File: ios/Airhop/AirhopIPtProxy.swift
// AirhopIPtProxy.swift
//
// The pluggable transports Arti dials to reach a bridge, built from
// native/iptproxy.
//
// Arti would normally run obfs4 and Snowflake as child processes, which iOS
// forbids. IPtProxy runs them in-process and exposes each as a SOCKS5 listener
// on loopback; this owns their lifecycle and reports the ports they landed on.
//
// Not main-actor isolated, and deliberately so: starting a transport launches a
// listener and, for Snowflake, contacts a broker over the network. Callers run
// it off the main thread, the same way they do Arti's own start.
//
// Mirrors android .../tor/AirhopIPtProxy.kt.

import Foundation
import IPtProxy
import OSLog

enum AirhopTransport: String {
    case obfs4
    case snowflake

    /// The transports named by a newline-separated list of bridge lines.
    ///
    /// Reads the first word of each line, where the Tor bridge format puts the
    /// transport name. Not a parser: Arti parses the lines authoritatively in
    /// `start`, so a line this misreads is refused there because its transport's
    /// port will be 0.
    static func named(in bridgeLines: String) -> Set<AirhopTransport> {
        var found: Set<AirhopTransport> = []
        for line in bridgeLines.split(separator: "\n") {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            guard !trimmed.isEmpty, !trimmed.hasPrefix("#") else { continue }
            var words = trimmed.split(separator: " ", omittingEmptySubsequences: true)
            // The word `Bridge` may lead the line; the transport follows it.
            if words.first?.lowercased() == "bridge" { words.removeFirst() }
            if let first = words.first, let transport = AirhopTransport(rawValue: String(first)) {
                found.insert(transport)
            }
        }
        return found
    }
}

final class AirhopIPtProxy: @unchecked Sendable {
    static let shared = AirhopIPtProxy()

    /// Guards both fields below. Every entry point is callable from any thread,
    /// and start and stop must not interleave.
    private let lock = NSLock()

    /// Nil until something asks for a transport: the controller writes to its
    /// state directory on construction, so it is not built for a user who never
    /// turns bridges on.
    private var controller: IPtProxyController?
    private var running: Set<AirhopTransport> = []

    private init() {}

    /// Where the transports keep state.
    ///
    /// Beside Arti's directory rather than inside it: the panic wipe deletes
    /// both, and Arti owns the layout of its own.
    static func stateDirectoryURL() -> URL? {
        guard let base = try? FileManager.default.url(
            for: .applicationSupportDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        ) else { return nil }
        return base.appendingPathComponent("airhop/iptproxy", isDirectory: true)
    }

    /// Start `transports`, returning the loopback port each is listening on, or
    /// nil if any could not start.
    ///
    /// A nil is fatal to the caller: Arti would otherwise be asked for a bridge
    /// whose transport is missing, and a user who asked for a bridge is likely
    /// somewhere a direct connection is unsafe.
    func start(_ transports: Set<AirhopTransport>) -> [AirhopTransport: UInt16]? {
        guard !transports.isEmpty else { return [:] }

        lock.lock()
        defer { lock.unlock() }

        guard let controller = ensureControllerLocked() else { return nil }

        var ports: [AirhopTransport: UInt16] = [:]
        for transport in transports {
            if !running.contains(transport) {
                do {
                    // No upstream proxy: Arti is the only thing dialling these.
                    try controller.start(transport.rawValue, proxy: "")
                } catch {
                    AirhopLog.iptProxy.error("\(transport.rawValue, privacy: .public) did not start")
                    stopLocked()
                    return nil
                }
                running.insert(transport)
            }
            // Ports are assigned by the library, never chosen. A zero means the
            // listener is not up despite start having returned.
            let port = controller.port(transport.rawValue)
            guard port > 0, port <= 65535 else {
                AirhopLog.iptProxy.error("\(transport.rawValue, privacy: .public) reported no port")
                stopLocked()
                return nil
            }
            AirhopLog.iptProxy.notice("\(transport.rawValue, privacy: .public) reported port \(port, privacy: .public)")
            ports[transport] = UInt16(port)
        }
        return ports
    }

    /// Stop every running transport. Safe to call when none are.
    func stop() {
        lock.lock()
        defer { lock.unlock() }
        stopLocked()
    }

    /// Stop the transports and delete everything they wrote.
    ///
    /// Panic wipe only. The state directory records which transports ran, and a
    /// gesture that promises local state is gone cannot leave that behind.
    func wipeState() {
        lock.lock()
        stopLocked()
        controller = nil
        lock.unlock()

        guard let dir = Self.stateDirectoryURL() else { return }
        try? FileManager.default.removeItem(at: dir)
    }

    // MARK: - Private

    private func stopLocked() {
        guard let controller else { return }
        for transport in running {
            controller.stop(transport.rawValue)
        }
        running.removeAll()
    }

    private func ensureControllerLocked() -> IPtProxyController? {
        if let controller { return controller }
        guard let dir = Self.stateDirectoryURL()?.path else { return nil }
        try? FileManager.default.createDirectory(
            at: URL(fileURLWithPath: dir), withIntermediateDirectories: true
        )

        // Logging off. The log records bridge addresses, which is the most
        // incriminating thing this device could write down, and nothing reads it.
        //
        // The initialiser returns nil rather than throwing when it cannot set up
        // its directory, so a nil here is a real failure.
        guard let created = IPtProxyController(
            dir,
            enableLogging: false,
            unsafeLogging: false,
            logLevel: "ERROR",
            transportEvents: nil
        ) else { return nil }


        controller = created
        return created
    }
}

